testcases/kernel/syscalls/fcntl/fcntl27.c

```diff
@@ -1,192 +1,51 @@
-     * parse standard options
-     ***************************************************************/
-	tst_parse_opts(ac, av, NULL, NULL);
-
-    /***************************************************************
-     * perform global setup for test
-     ***************************************************************/
-	setup();
-
-	expected_result = -1;
-
-    /***************************************************************
-     * check looping state if -c option given
-     ***************************************************************/
-	for (lc = 0; TEST_LOOPING(lc); lc++) {
+ * Basic test for fcntl(2) using F_SETLEASE & F_RDLCK argument.
+ */
 
-		tst_count = 0;
+#include "lapi/fcntl.h"
+#include "tst_test.h"
 
 #ifdef F_SETLEASE
-		/*
-		 * Call fcntl(2) with F_SETLEASE & F_RDLCK argument on fname
-		 */
-		TEST(fcntl(fd, F_SETLEASE, F_RDLCK));
-
-		/* check return code */
-		if (TEST_RETURN == expected_result) {
-			tst_resm(TPASS,
-				 "fcntl(fd, F_SETLEASE, F_RDLCK) succeeded");
-		} else {
-			tst_resm(TFAIL, "fcntl(%s, F_SETLEASE, F_RDLCK)"
-				 " failed with errno %d : %s", fname,
-				 TEST_ERRNO, strerror(TEST_ERRNO));
-		}
-#else
-		tst_resm(TINFO, "F_SETLEASE not defined, skipping test");
-#endif
-	}
 
-    /***************************************************************
-     * cleanup and exit
-     ***************************************************************/
-	cleanup();
-	tst_exit();
-
-}
-
-/***************************************************************
- * setup() - performs all ONE TIME setup for this test.
- ***************************************************************/
-void setup(void)
+static struct test_case
 {
-
-	tst_sig(NOFORK, DEF_HANDLER, cleanup);
-
-	TEST_PAUSE;
-
-	tst_tmpdir();
-
-	sprintf(fname, "tfile_%d", getpid());
-	if ((fd = open(fname, O_RDWR | O_CREAT, 0777)) == -1) {
-		tst_brkm(TBROK, cleanup,
-			 "open(%s, O_RDWR|O_CREAT,0777) Failed, errno=%d : %s",
-			 fname, errno, strerror(errno));
-	}
-}
-
-/***************************************************************
- * cleanup() - performs all ONE TIME cleanup for this test at
- *				 completion or premature exit.
- ***************************************************************/
-void cleanup(void)
+	int oflags;
+	mode_t mode;
+	int cmd;
+	char *cmd_desc;
+	int arg;
+	char *arg_desc;
+	int exp_errno;
+} tcases[] = {
+	{O_RDWR | O_CREAT, 0777, F_SETLEASE, "F_SETLEASE", F_RDLCK, "F_RDLCK", EAGAIN},
+	{O_WRONLY | O_CREAT, 0222, F_SETLEASE, "F_SETLEASE", F_RDLCK, "F_RDLCK", EAGAIN},
+};
+
+static void verify_fcntl(unsigned int nr)
 {
+	struct test_case *tc = &tcases[nr];
+	int fd = SAFE_OPEN("testfile", tc->oflags, tc->mode);
 
-	/* close the file we've had open */
-	if (close(fd) == -1) {
-		tst_resm(TWARN, "close(%s) Failed, errno=%d : %s", fname, errno,
-			 strerror(errno));
-	}
+	TST_EXP_FAIL(fcntl(fd, tc->cmd, tc->arg), tc->exp_errno, "fcntl(fd, %s, %s)", tc->cmd_desc, tc->arg_desc);
```

Comment:
In this case I would use it without optional config (`TST_EXP_FAIL(fcntl(fd, tc->cmd, tc->arg), tc->exp_errno)`, because `cmd` and `arg` are constants, which are printed anyway by `TST_EXP_FAIL()` without args. The only difference would be with `fd`, but you don't print `fd` anyway.

Reply:
I'm sorry, I was wrong here. While constants are passed to macro would be printed, e.g.

```C
TST_EXP_FAIL(fcntl(fd, F_SETLEASE, F_RDLCK), EAGAIN);
```
prints
```
fcntl27.c:41: TPASS: fcntl(fd, F_SETLEASE, F_RDLCK) : EAGAIN/EWOULDBLOCK (11)
```
using `TST_EXP_FAIL(fcntl(fd, tc->cmd, tc->arg), EAGAIN);` obviously does not print macros. Therefore optional string to describe what is passed is probably better.